File: services/ai/app/tools.py
"""Agent tools — mirror lib/ai/tools.ts. Same env var names + defaults so TS and
Python configuration stay identical."""

import os

OWNER_EMAIL = "naingoted@gmail.com"


def get_contact_info(
    email: str = OWNER_EMAIL, github: str | None = None
) -> dict[str, str]:
    info = {"email": email}
    gh = github if github is not None else os.getenv("NEXT_PUBLIC_GITHUB_URL")
    if gh:
        info["github"] = gh
    return info


def schedule_call(url: str | None = None) -> dict[str, str]:
    return {"url": url or os.getenv("NEXT_PUBLIC_CAL_URL", "https://cal.com/tet-nai")}


def show_resume() -> dict[str, str]:
    return {
        "path": "/resume",
        "description": (
            "The resume is available for download at /resume. Offer this link to "
            "the visitor."
        ),
    }


def how_this_works() -> str:
    return (
        "meclaw is Thet Naing's personal bot. Phase 3 runs a Python FastAPI sidecar whose "
        "core is a LangGraph state graph (triage -> specialized agent -> review -> "
        "respond). Knowledge comes from markdown in content/, embedded with Ollama "
        "nomic-embed-text and retrieved from Qdrant. The chat model is qwen3.6-plus "
        "via an Anthropic-compatible gateway. The Next.js app handles the UI, "
        "guardrails, and Postgres persistence."
    )

File: lib/ai/tools.ts
import { tool } from "ai";
import { z } from "zod";

/**
 * Owner's email address — single source of truth.
 * Used by getContactInfo tool. Update this if the owner's email changes.
 */
const OWNER_EMAIL = "thetnaing@incube8.sg";

/**
 * getContactInfo — returns the owner's contact details.
 * The model can call this when asked "How do I get in touch?" or similar.
 */
export const getContactInfo = tool({
  description:
    "Get the owner's contact information including email and GitHub profile (if available).",
  inputSchema: z.object({}),
  execute: () => {
    const contactInfo: Record<string, string> = {
      email: OWNER_EMAIL,
    };

    // Include GitHub URL if set in environment
    if (process.env.NEXT_PUBLIC_GITHUB_URL) {
      contactInfo.github = process.env.NEXT_PUBLIC_GITHUB_URL;
    }

    return contactInfo;
  },
});

/**
 * scheduleCall — returns the Cal.com booking link.
 * The model can call this when asked "Can I schedule a call?" or "Let's book a meeting".
 */
export const scheduleCall = tool({
  description:
    "Get the Cal.com booking link so the visitor can schedule a call with the owner.",
  inputSchema: z.object({}),
  execute: () => {
    const calUrl = process.env.NEXT_PUBLIC_CAL_URL || "https://cal.com/tet-nai";
    return {
      url: calUrl,
    };
  },
});

/**
 * showResume — returns the resume download path and instructions.
 * The model can call this when asked "Can I see the resume?" or "Download CV".
 */
export const showResume = tool({
  description:
    "Get the resume download link. Returns a path to the resume file.",
  inputSchema: z.object({}),
  execute: () => {
    return {
      path: "/resume",
      description:
        "The resume is available for download at /resume. You can offer this link to the visitor.",
    };
  },
});

/**
 * howThisWorks — meta tool that explains the bot's architecture.
 * The model can call this when asked "How does this bot work?" or "Tell me about yourself".
 */
export const howThisWorks = tool({
  description:
    "Explain how this bot works: its architecture, knowledge source, and tech stack.",
  inputSchema: z.object({}),
  execute: () => {
    return `echo is a personal AI twin built with Next.js 16 and the Vercel AI SDK v6 (streaming).
The LLM is qwen3.6-plus, accessed via an Anthropic-compatible gateway.
Knowledge about Thet comes from markdown files in the content/ directory. These are chunked and
embedded (nomic-embed-text via Ollama) into a Qdrant vector database; each question is embedded and
the most relevant chunks are retrieved and added to the system prompt (retrieval-augmented generation).
If retrieval is unavailable, it falls back to stuffing the full corpus into the prompt. Conversations
are persisted to a local SQLite database (better-sqlite3) for future reference. The goal is to create
a personalized, always-available chatbot that answers about the owner in a warm, third-person voice.`;
  },
});

/**
 * Tools registry — export all tools as a record for wiring into streamText.
 * The model can decide which tools to call based on user intent.
 */
export const tools = {
  getContactInfo,
  scheduleCall,
  showResume,
  howThisWorks,
};
